File: backend/models/LabTest.js
/**
 * ═══════════════════════════════════════════════════════════════════════════
 *  LabTest Model — Patient 360°
 *  ─────────────────────────────────────────────────────────────────────────
 *  Collection: lab_tests
 *  Source of truth: patient360_db_final.js (collection 17)
 *
 *  Lab test orders and results. Lifecycle:
 *
 *    1. ordered           — doctor creates the order from a Visit
 *    2. scheduled         — patient books an appointment at the lab
 *    3. sample_collected  — lab tech collects the sample, prints barcode
 *    4. in_progress       — sample is being processed
 *    5. completed         — results entered + PDF uploaded; both view flags
 *                           reset to false to trigger notifications
 *    6. (terminal)        — cancelled / rejected
 *
 *  Result data:
 *    • testResults[]   — structured key/value results for each individual test
 *                        (e.g. CBC has WBC, RBC, Hgb, Plt as separate items)
 *                        Each has reference range, isAbnormal, isCritical
 *    • resultPdfUrl    — the official PDF report from the lab
 *
 *  Critical results (isCritical=true on any line item) trigger the
 *  lab_results_critical notification type with priority='urgent' so the
 *  ordering doctor is alerted immediately.
 *
 *  View tracking (isViewedByDoctor / isViewedByPatient) lets the
 *  notification dispatcher know who still needs to be pinged.
 * ═══════════════════════════════════════════════════════════════════════════
 */

const mongoose = require('mongoose');

const { Schema } = mongoose;

// ── Enums ───────────────────────────────────────────────────────────────────

const TEST_CATEGORIES = [
  'blood', 'urine', 'stool', 'imaging', 'biopsy',
  'microbiology', 'molecular', 'other',
];

const PRIORITIES = ['routine', 'urgent', 'stat'];

const SAMPLE_TYPES = [
  'blood', 'urine', 'stool', 'tissue', 'swab', 'saliva', 'other',
];

const STATUSES = [
  'ordered', 'scheduled', 'sample_collected',
  'in_progress', 'completed', 'cancelled', 'rejected',
];

const CURRENCIES = ['SYP', 'USD'];

// ── Sub-schemas ─────────────────────────────────────────────────────────────

const TestOrderedSchema = new Schema(
  {
    testCode: {
      type: String,
      required: true,
      trim: true,
      uppercase: true,
    },
    testName: { type: String, required: true, trim: true },
    notes: { type: String, trim: true },
  },
  { _id: false },
);

const TestResultSchema = new Schema(
  {
    testCode: { type: String, trim: true, uppercase: true },
    testName: {
      type: String,
      required: [true, 'اسم الاختبار مطلوب'],
      trim: true,
    },
    value: {
      type: String,
      required: [true, 'قيمة النتيجة مطلوبة'],
      trim: true,
      // Stored as string to support qualitative results ("Positive", "Trace")
      // alongside numeric values
    },
    numericValue: { type: Number },     // for trend/graph plotting
    unit: { type: String, trim: true }, // e.g. mg/dL, g/L, %
    referenceRange: { type: String, trim: true }, // e.g. "70-100"
    isAbnormal: { type: Boolean, default: false },
    isCritical: { type: Boolean, default: false },
  },
  { _id: false },
);

// ── Main schema ──────────────────────────────────────────────────────────────

const LabTestSchema = new Schema(
  {
    // ── Auto-generated identifier ─────────────────────────────────────────
    testNumber: {
      type: String,
      unique: true,
      trim: true,
      uppercase: true,
      // Format: LAB-YYYYMMDD-XXXXX
    },

    // ── Patient (XOR) ─────────────────────────────────────────────────────
    patientPersonId: {
      type: Schema.Types.ObjectId,
      ref: 'Person',
      sparse: true,
    },
    patientChildId: {
      type: Schema.Types.ObjectId,
      ref: 'Children',
      sparse: true,
    },

    // ── Ordering doctor + originating visit ───────────────────────────────
    orderedBy: {
      type: Schema.Types.ObjectId,
      ref: 'Doctor',
      required: [true, 'الطبيب الطالب مطلوب'],
    },
    visitId: { type: Schema.Types.ObjectId, ref: 'Visit', sparse: true },

    // ── Performing laboratory ─────────────────────────────────────────────
    laboratoryId: {
      type: Schema.Types.ObjectId,
      ref: 'Laboratory',
      required: [true, 'المختبر مطلوب'],
    },

    // ── Order metadata ────────────────────────────────────────────────────
    orderDate: { type: Date, default: Date.now, required: true },
    scheduledDate: { type: Date },

    // ── Tests requested ───────────────────────────────────────────────────
    testsOrdered: {
      type: [TestOrderedSchema],
      validate: {
        validator: (arr) => Array.isArray(arr) && arr.length >= 1,
        message: 'يجب طلب اختبار واحد على الأقل',
      },
    },

    testCategory: { type: String, enum: TEST_CATEGORIES },
    priority: { type: String, enum: PRIORITIES, default: 'routine' },

    // ── Sample collection ─────────────────────────────────────────────────
    sampleType: { type: String, enum: SAMPLE_TYPES },
    sampleId: { type: String, trim: true, sparse: true }, // lab barcode
    sampleCollectedAt: { type: Date },
    sampleCollectedBy: {
      type: Schema.Types.ObjectId,
      ref: 'LabTechnician',
    },

    // ── Status ────────────────────────────────────────────────────────────
    status: {
      type: String,
      enum: STATUSES,
      default: 'ordered',
      index: true,
    },
    rejectionReason: { type: String, trim: true },

    // ── Results ───────────────────────────────────────────────────────────
    testResults: { type: [TestResultSchema], default: [] },

    resultPdfUrl: { type: String, trim: true },
    resultPdfUploadedAt: { type: Date },
    resultPdfUploadedBy: {
      type: Schema.Types.ObjectId,
      ref: 'LabTechnician',
    },

    labNotes: { type: String, trim: true },
    completedAt: { type: Date },
    completedBy: {
      type: Schema.Types.ObjectId,
      ref: 'LabTechnician',
    },

    // ── View tracking (drives notifications) ──────────────────────────────
    isCritical: { type: Boolean, default: false },
    isViewedByDoctor: { type: Boolean, default: false },
    isViewedByPatient: { type: Boolean, default: false },
    doctorViewedAt: { type: Date },
    patientViewedAt: { type: Date },

    // ── Pricing ───────────────────────────────────────────────────────────
    totalCost: { type: Number, min: 0 },
    currency: { type: String, enum: CURRENCIES, default: 'SYP' },
  },
  {
    timestamps: true,
    collection: 'lab_tests',
    toJSON: { virtuals: true },
    toObject: { virtuals: true },
  },
);

// ── Indexes ─────────────────────────────────────────────────────────────────

LabTestSchema.index(
  { patientPersonId: 1, orderDate: -1 },
  { name: 'idx_patient_adult_date' },
);
LabTestSchema.index(
  { patientChildId: 1, orderDate: -1 },
  { name: 'idx_patient_child_date' },
);
LabTestSchema.index(
  { orderedBy: 1, orderDate: -1 },
  { name: 'idx_doctor_date' },
);
LabTestSchema.index(
  { laboratoryId: 1, status: 1 },
  { name: 'idx_lab_status' },
);
LabTestSchema.index(
  { status: 1, scheduledDate: 1 },
  { name: 'idx_status_scheduled' },
);
LabTestSchema.index({ isCritical: 1 }, { name: 'idx_critical' });
LabTestSchema.index({ isViewedByDoctor: 1 }, { name: 'idx_viewedByDoctor' });
LabTestSchema.index({ isViewedByPatient: 1 }, { name: 'idx_viewedByPatient' });
LabTestSchema.index({ visitId: 1 }, { sparse: true, name: 'idx_visitId' });

// ── Static helpers ──────────────────────────────────────────────────────────

/**
 * Generate test number: LAB-YYYYMMDD-XXXXX (scoped per day).
 * @returns {Promise<string>}
 */
LabTestSchema.statics.generateTestNumber = async function generateTestNumber() {
  const now = new Date();
  const yyyy = now.getFullYear();
  const mm = String(now.getMonth() + 1).padStart(2, '0');
  const dd = String(now.getDate()).padStart(2, '0');
  const datePart = `${yyyy}${mm}${dd}`;
  const prefix = `LAB-${datePart}-`;

  const todayCount = await this.countDocuments({
    testNumber: { $regex: `^${prefix}` },
  });

  const sequence = String(todayCount + 1).padStart(5, '0');
  return `${prefix}${sequence}`;
};

// ── Pre-validate: patient XOR ───────────────────────────────────────────────

LabTestSchema.pre('validate', function enforcePatientXor(next) {
  const hasPerson = !!this.patientPersonId;
  const hasChild = !!this.patientChildId;
  if (!hasPerson && !hasChild) {
    return next(new Error('يجب تحديد patientPersonId أو patientChildId'));
  }
  if (hasPerson && hasChild) {
    return next(new Error('لا يمكن تحديد patientPersonId و patientChildId معاً'));
  }
  return next();
});

// ── Pre-save: auto-generate testNumber + compute isCritical ────────────────

LabTestSchema.pre('save', async function autoFields(next) {
  try {
    if (this.isNew && !this.testNumber) {
      this.testNumber = await this.constructor.generateTestNumber();
    }

    // Auto-set isCritical at the test-level if any result is critical
    if (this.isModified('testResults')) {
      this.isCritical = (this.testResults || []).some((r) => r.isCritical);
    }

    return next();
  } catch (err) {
    return next(err);
  }
});

// ── Virtuals ────────────────────────────────────────────────────────────────

LabTestSchema.virtual('hasResults').get(function () {
  return Array.isArray(this.testResults) && this.testResults.length > 0;
});

LabTestSchema.virtual('hasPdfReport').get(function () {
  return !!this.resultPdfUrl;
});

LabTestSchema.virtual('abnormalCount').get(function () {
  if (!Array.isArray(this.testResults)) return 0;
  return this.testResults.filter((r) => r.isAbnormal).length;
});

// ── Instance methods ────────────────────────────────────────────────────────

/**
 * Mark this test as viewed by a specific viewer type, stamping the time.
 * Called when doctor or patient opens the result.
 *
 * @param {'doctor' | 'patient'} viewerType
 */
LabTestSchema.methods.markViewedBy = async function markViewedBy(viewerType) {
  if (viewerType === 'doctor') {
    this.isViewedByDoctor = true;
    this.doctorViewedAt = new Date();
  } else if (viewerType === 'patient') {
    this.isViewedByPatient = true;
    this.patientViewedAt = new Date();
  } else {
    throw new Error('viewerType must be "doctor" or "patient"');
  }
  return this.save();
};

/**
 * Transition to 'completed' state. Stamps completion metadata and resets
 * view flags so notifications fire.
 *
 * @param {ObjectId} labTechnicianId
 */
LabTestSchema.methods.markCompleted = async function markCompleted(labTechnicianId) {
  this.status = 'completed';
  this.completedAt = new Date();
  this.completedBy = labTechnicianId;
  // Reset view flags so notifications dispatch to both parties
  this.isViewedByDoctor = false;
  this.isViewedByPatient = false;
  return this.save();
};

module.exports = mongoose.model('LabTest', LabTestSchema);

File: patient360_mobile/lib/features/lab_results/domain/lab_test.dart
import 'package:flutter/foundation.dart';

import 'test_ordered.dart';
import 'test_result_row.dart';

/// Mirrors the `lab_tests` collection. Read-only from the mobile app —
/// the patient marks a completed test viewed but never modifies its
/// results, status, or attached PDF.
@immutable
class LabTest {
  const LabTest({
    required this.id,
    required this.testNumber,
    required this.orderDate,
    required this.testCategory,
    required this.priority,
    required this.status,
    required this.testsOrdered,
    required this.testResults,
    required this.isCritical,
    required this.isViewedByPatient,
    required this.createdAt,
    this.patientPersonId,
    this.patientChildId,
    this.orderedBy,
    this.visitId,
    this.laboratoryId,
    this.scheduledDate,
    this.sampleType,
    this.sampleId,
    this.sampleCollectedAt,
    this.rejectionReason,
    this.resultPdfUrl,
    this.resultPdfUploadedAt,
    this.labNotes,
    this.completedAt,
    this.patientViewedAt,
    this.totalCost,
  });

  factory LabTest.fromJson(Map<String, dynamic> json) {
    DateTime asDate(Object? v, {DateTime? fallback}) {
      if (v is String && v.isNotEmpty) return DateTime.parse(v);
      return fallback ?? DateTime.fromMillisecondsSinceEpoch(0);
    }

    DateTime? asDateOrNull(Object? v) =>
        v is String && v.isNotEmpty ? DateTime.parse(v) : null;

    final List<TestOrdered> testsOrdered =
        (json['testsOrdered'] as List<dynamic>?)
                ?.whereType<Map<dynamic, dynamic>>()
                .map((Map<dynamic, dynamic> m) => m.cast<String, dynamic>())
                .map(TestOrdered.fromJson)
                .toList() ??
            const <TestOrdered>[];

    final List<TestResultRow> testResults =
        (json['testResults'] as List<dynamic>?)
                ?.whereType<Map<dynamic, dynamic>>()
                .map((Map<dynamic, dynamic> m) => m.cast<String, dynamic>())
                .map(TestResultRow.fromJson)
                .toList() ??
            const <TestResultRow>[];

    return LabTest(
      id: (json['_id'] ?? json['id']).toString(),
      testNumber: (json['testNumber'] as String?) ?? '',
      patientPersonId: json['patientPersonId'] as String?,
      patientChildId: json['patientChildId'] as String?,
      orderedBy: json['orderedBy'] as String?,
      visitId: json['visitId'] as String?,
      laboratoryId: json['laboratoryId'] as String?,
      orderDate: asDate(json['orderDate']),
      scheduledDate: asDateOrNull(json['scheduledDate']),
      testsOrdered: testsOrdered,
      testCategory: (json['testCategory'] as String?) ?? '',
      priority: (json['priority'] as String?) ?? 'routine',
      sampleType: json['sampleType'] as String?,
      sampleId: json['sampleId'] as String?,
      sampleCollectedAt: asDateOrNull(json['sampleCollectedAt']),
      status: (json['status'] as String?) ?? 'ordered',
      rejectionReason: json['rejectionReason'] as String?,
      testResults: testResults,
      resultPdfUrl: json['resultPdfUrl'] as String?,
      resultPdfUploadedAt: asDateOrNull(json['resultPdfUploadedAt']),
      labNotes: json['labNotes'] as String?,
      completedAt: asDateOrNull(json['completedAt']),
      isCritical: (json['isCritical'] as bool?) ?? false,
      isViewedByPatient: (json['isViewedByPatient'] as bool?) ?? false,
      patientViewedAt: asDateOrNull(json['patientViewedAt']),
      totalCost: (json['totalCost'] as num?)?.toDouble(),
      createdAt: asDate(json['createdAt'], fallback: DateTime.now()),
    );
  }

  final String id;
  final String testNumber;
  final String? patientPersonId;
  final String? patientChildId;
  final String? orderedBy;
  final String? visitId;
  final String? laboratoryId;
  final DateTime orderDate;
  final DateTime? scheduledDate;
  final List<TestOrdered> testsOrdered;
  final String testCategory;
  final String priority;
  final String? sampleType;
  final String? sampleId;
  final DateTime? sampleCollectedAt;

  /// One of: `ordered | scheduled | sample_collected | in_progress |
  /// completed | cancelled | rejected`.
  final String status;
  final String? rejectionReason;
  final List<TestResultRow> testResults;
  final String? resultPdfUrl;
  final DateTime? resultPdfUploadedAt;
  final String? labNotes;
  final DateTime? completedAt;
  final bool isCritical;
  final bool isViewedByPatient;
  final DateTime? patientViewedAt;
  final double? totalCost;
  final DateTime createdAt;

  bool get isCompleted => status == 'completed';
  bool get hasResults => testResults.isNotEmpty;

  /// Count of result rows flagged abnormal (includes critical — critical
  /// is a *severity-amplified* form of abnormal in the schema).
  int get abnormalCount =>
      testResults.where((TestResultRow r) => r.isAbnormal || r.isCritical).length;

  /// Count of result rows flagged critical specifically. Drives the
  /// persistent SnackBar reminder on first expand.
  int get criticalCount =>
      testResults.where((TestResultRow r) => r.isCritical).length;

  LabTest copyWith({
    bool? isViewedByPatient,
    DateTime? patientViewedAt,
  }) {
    return LabTest(
      id: id,
      testNumber: testNumber,
      patientPersonId: patientPersonId,
      patientChildId: patientChildId,
      orderedBy: orderedBy,
      visitId: visitId,
      laboratoryId: laboratoryId,
      orderDate: orderDate,
      scheduledDate: scheduledDate,
      testsOrdered: testsOrdered,
      testCategory: testCategory,
      priority: priority,
      sampleType: sampleType,
      sampleId: sampleId,
      sampleCollectedAt: sampleCollectedAt,
      status: status,
      rejectionReason: rejectionReason,
      testResults: testResults,
      resultPdfUrl: resultPdfUrl,
      resultPdfUploadedAt: resultPdfUploadedAt,
      labNotes: labNotes,
      completedAt: completedAt,
      isCritical: isCritical,
      isViewedByPatient: isViewedByPatient ?? this.isViewedByPatient,
      patientViewedAt: patientViewedAt ?? this.patientViewedAt,
      totalCost: totalCost,
      createdAt: createdAt,
    );
  }
}

/// UI bucket for the 3-tab filter on the lab results screen.
enum LabTestGroup { all, pending, completed }

extension LabTestGrouping on LabTestGroup {
  bool includes(LabTest t) {
    switch (this) {
      case LabTestGroup.all:
        return true;
      case LabTestGroup.pending:
        return t.status != 'completed';
      case LabTestGroup.completed:
        return t.status == 'completed';
    }
  }
}
